readthedocs/projects/models.py

```diff
@@ -225,6 +230,35 @@ class Project(models.Model):
         ),
     )
 
+    urlpattern = models.CharField(
+        _("Custom URL pattern for multi version and single version projects"),
+        max_length=255,
+        default=None,
+        blank=True,
+        null=True,
+        help_text=_(
+            "A Python regex pattern used when evaluating a multi version or single version project. "  # noqa
+            "For multi version projects it needs to declare the following replacement fields: language, version, and filename. "  # noqa
+            "For single version projects it needs to declare the filename replacement field only. "
+            "The default pattern for multi version projects is: `^/{language}(/({version}(/{filename})?)?)?$`, "  # noqa
+            "and for single version projects is: `^/{filename}?$`."
+        ),
+    )
+    urlpattern_subproject = models.CharField(
```

Comment:
I'm also curious why this is required. I'm sure you know, but it would be useful to write down somewhere. At least in a comment in this file, if not a full design doc :)

Reply:
Some of this is described at https://github.com/readthedocs/readthedocs.org/blob/7e191bb4e674863c775d3927b54ef6813efc086f/readthedocs/proxito/README.rst?plain=1#L72-L77

The old implementation was trying to fit the pattern for subprojects and normal projects in one field/pattern, but they need to be separate (https://github.com/readthedocs/readthedocs.org/pull/8327).